Infrastructure Design Proposal

jbkolze opened the pull request:
I’ve had a little time to fiddle with / brainstorm on the event manager this week.  Below is a diagram of what I’m currently envisioning as far as infrastructure setup goes.  Open to any thoughts / suggestions / complaints before I get too deep in the weeds.

![wm-event-manager infrastructure diagram](https://raw.githubusercontent.com/USACE-WaterManagement/wm-event-manager/19ebf67156463b7aac52bf013b6b2e04530a4859/proposals/wm-event-manager.drawio.png)

I think this particular use case (ad hoc / event-driven) would lend itself well to a serverless architecture, so I included those options where feasible (split by vertical line).  There's definitely a discussion to be had there regarding cost, familiarity with the architecture, etc.  I currently have DynamoDB implemented locally and it was a pretty easy setup after some initial research on design best practices -- seems like a good fit for a relatively small DB instance.

Feel free to dump thoughts here or we can set up a meeting to discuss.  I'll continue plodding along as time allows in the meantime.

jbkolze committed:
Add proposed infrastructure diagrams

jbkolze commented:
Thanks for the input, Mike.  Jotting some notes here for reference later:
3. Agree -- I think this use case is simple enough that I have a decent handle on schema after a little research: Table PK of JobId with GSIs to query by User, Office, etc.  Hopefully any issues will pop up in dev if we need to pivot.  I certainly welcome review by someone with DDB experience if we know of anyone, though.
4. AWS automatically manages SQS polling from Lambda in the background, so I think that would be the way to go in this case unless we just had some philosophical resistance to Lambda.
5. This might be worth some discussion because Batch is already orchestrating ECS to some degree, and SQS -> Lambda -> Batch is going to add some overhead (both on the code and infrastructure sides) with potentially not much gain for our uses.  Just don't want to prematurely optimize.
6. Agree that this probably makes the most sense.
7. I think there are a few ways to handle this (polling the Batch job status, creating an EventBridge rule) that just depend on how much we want to lean on AWS infra.  Probably leads into a bigger conversation re: balancing convenience, vendor lock-in, external pressure, etc.

MikeNeilson commented:
4. Opposite, I've been trying to find a reason to use it. Partly to get the experience but also because it often seems like a reasonable method.. and especially so in this case.
5. Yeah, that's always a challenge
7. Using "messages" we aren't tied to AWS, just the concept of messaging, so it's reasonable to assume they'll be used and that at the moment it's SQS. The applications don't, and shouldn't, be written in a way that assumes SQS, just "messaging" but that's easier said than done in different languages.

Java has JMS for which SQS can just be another provider, I'm not aware of a similar mechanism in Python but the concept of Messaging is well defined enough we can create our own interface/wrapper over time if we don't find one.

jbkolze committed:
Update diagram with current design decisions

jbkolze committed:
Move design diagrams to docs/

jbkolze commented:
Latest proposed infrastructure diagram update:
![wm-event-manager infrastructure diagram](https://raw.githubusercontent.com/USACE-WaterManagement/wm-event-manager/c62de3839061243662ccc2c0b67877c58190303f/docs/wm-event-manager.drawio.png)

jbkolze committed:
Update design diagram based on comments

- Adjust S3 section to include job catalogs
- Clean up keycloak/CDA connectivity
- Re-enable SQS job queue and dispatcher in front of job runners

MikeNeilson commented:
> You're right that there's an arrow missing -- there should be an arrow from the front end to keycloak. The front end gets a key from keycloak (using CAC), but the API does independently verify that key with keycloak (for security) and uses the CDA /user/profile endpoint to determine which offices the user can view/run jobs for. As you mentioned, the permissions model is pretty sparse right now so it just verifies the user has the "CWMS User" role.
> 
> edit: fixed some CDA/keycloak language

This seems reasonable for now. To go beyond "user has a simple role" we will likely need to reach out to the OPA agent when it exists. Good news is, that really just replaces that particular arrow; since you're already making an HTTP call (presumably on the backend for actual security... never trust the client, of course) we shouldn't even need an infrastructure change when the time comes.

jbkolze committed:
Update diagram with latest decisions

- Use ECS for the API
- Pull job logs directly from CloudWatch
- Use RDS/PostgreSQL for database
- Insert SQS job queue & lambda dispatcher in front of runner

jbkolze committed:
Update diagram for current implementation
